functions/access.js

```diff
@@ -0,0 +1,136 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!venueId || !uid || !token) return false;
+
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  return await admin
+    .firestore()
+    .runTransaction(async (transaction) => {
+      const venue = await transaction.get(venueRef);
+      const granted = await transaction.get(accessRef);
+
+      if (!venue.exists) {
+        return false;
+      }
+      if (!granted.exists || !granted.tokens) {
+        return false;
+      }
+
+      if (Object.keys(granted.tokens).includes(token)) {
+        // Record that the token was checked
+        if (!granted.tokens[token].usedAt) {
+          granted.tokens[token].usedAt = [];
+        }
+        granted.tokens[token].usedAt.push(Date.now());
+        transaction.update(accessRef, granted);
+
+        return true;
+      }
+
+      return false;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
+      return false;
+    });
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  if (!venueId || !uid || !password || !email || !code) return {};
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+
+  const token = uuid();
+  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
```

Comment:
This is the same basic pattern as used in `checkIsValidToken`, where it can be refactored to use `venueRef`/`accessRef`, and then fetch them both in parallel using `Promise.all` (all inside a transaction using `.runTransaction`)

We can probably refactor the code slightly to share this common part between the two functions (and potentially any other places, if other places are doing similar)